File: frontend/src/api/contacts.ts
import { api } from "./client";

export type Contact = {
  id: string;
  name: Record<string, string>;
  title?: Record<string, string>;
  email?: string | null;
  phone?: string | null;
  sort_order?: number;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
};

export type ContactPayload = {
  name: Record<string, string>;
  title?: Record<string, string>;
  email?: string;
  phone?: string;
  sort_order?: number;
  is_active?: boolean;
};

export async function fetchContacts(): Promise<Contact[]> {
  return api("/contacts");
}

export async function fetchAdminContacts(all = true): Promise<Contact[]> {
  const query = all ? "?all=true" : "";
  return api(`/admin/contacts${query}`);
}

export async function createContact(payload: ContactPayload) {
  return api("/admin/contacts", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateContact(id: string, payload: ContactPayload) {
  return api(`/admin/contacts/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export async function deleteContact(id: string) {
  return api(`/admin/contacts/${id}`, { method: "DELETE" });
}

File: backend/internal/models/contacts.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
)

type Contact struct {
	ID        string       `db:"id" json:"id"`
	TenantID  string       `db:"tenant_id" json:"tenant_id"`
	Name      LocalizedMap `db:"name" json:"name"`
	Title     LocalizedMap `db:"title" json:"title,omitempty"`
	Email     *string      `db:"email" json:"email,omitempty"`
	Phone     *string      `db:"phone" json:"phone,omitempty"`
	SortOrder int          `db:"sort_order" json:"sort_order"`
	IsActive  bool         `db:"is_active" json:"is_active"`
	CreatedAt string       `db:"created_at" json:"created_at"`
	UpdatedAt string       `db:"updated_at" json:"updated_at"`
}

type LocalizedMap map[string]string

func (m *LocalizedMap) Scan(value interface{}) error {
	if value == nil {
		*m = LocalizedMap{}
		return nil
	}
	var b []byte
	switch v := value.(type) {
	case []byte:
		b = v
	case string:
		b = []byte(v)
	default:
		return fmt.Errorf("unsupported type %T", v)
	}
	if len(b) == 0 {
		*m = LocalizedMap{}
		return nil
	}
	return json.Unmarshal(b, m)
}

func (m LocalizedMap) Value() (driver.Value, error) {
	if len(m) == 0 {
		return nil, nil
	}
	return json.Marshal(m)
}
